pkg/ebpf/c/tcp-queue-length-kern.c

```diff
@@ -1,128 +1,101 @@
 #include <linux/kconfig.h>
 #define KBUILD_MODNAME "foo"
-#include <linux/ptrace.h>
 #include <linux/bpf.h>
-#include <net/inet_sock.h>
+#include <linux/ptrace.h>
 #include <linux/tcp.h>
+#include <net/inet_sock.h>
 
 #include "pkg/ebpf/c/bpf-common.h"
 #include "pkg/ebpf/tcp-queue-length-kern-user.h"
 
 /*
- * The `queue` map is used to share with the userland program system-probe
- * the statistics (max/min size of receive/send buffer) of every socket
+ * The `stats` map is used to share with the userland program system-probe
+ * the statistics (max size of receive/send buffer)
  */
-BPF_HASH(queue, struct sock *, struct stats);
+BPF_TABLE("percpu_hash", struct stats_key, struct stats_value, stats, 1024);
```

Comment:
What are your kernel version targets? This map type is not available until kernel v4.6.

Reply:
I'm guessing we need this so the `max` operation is thread safe? We could use atomic operations to get around this.

https://llvm.org/docs/Atomics.html#libcalls-atomic

Reply:
Indeed, that’s because the `max` computation needs to be thread safe.

I’m not sure atomic operations could help here because there is no atomic maximum so that we would have to implement it ourselves with an atomic compare and swap.
The use of atomic CAS requires to put it inside a loop and I would expect the eBPF byte-code verifier to reject it.

Reply:
Here is how I would implement a thread-safe max with atomic operations:
```c
#include <stdbool.h>

void max(int *max, int val)
{
	bool ret;
	int current_max;
	__atomic_load(max, &current_max, __ATOMIC_ACQUIRE);
	do {
		if (val <= current_max)
			return;
		ret = __atomic_compare_exchange_n(max, &current_max, val, false, __ATOMIC_RELEASE, __ATOMIC_ACQUIRE);
	} while(!ret);
}
```

And in fact, `clang` cannot even translate this into an eBPF bytecode:
`clang -o - -S -O cas.c` works and shows the x86 assembly code for the thread-safe max function. But
`clang -o - -S -O -target bpf cas.c` fails (at least with my `clang 9.0.0`)

Reply:
I suppose that the impossibility to build our own thread-safe functions with atomic operations is the reason why the `bpf_spin_(un)lock` eBPF helpers have been introduced.
But `bpf_spin_(un)lock` have been introduced in kernel 5.1 and their use causes a heavy performance penalty compared to “per cpu” data structures.

Reply:
About the portability concern, I have implemented a degraded fallback alternative in the python `network` check in DataDog/integrations-core#7861.